src/ipc/dbus/DBusObjectRegistry.cpp

```diff
@@ -78,44 +106,110 @@ void DBusObjectRegistry::unregisterObject(const QString &objectPath)
+{
+    int version = objectMap[versionKey].toInt(); // 0 if key is not present
+    if (version && (version > m_registryVersion ||
+            ((std::numeric_limits<int>::max() - m_registryVersion < 100) && (version < 100)))) // version number overflow
+    {
+        m_registryVersion = version;
+        m_objects = objectMap;
+        m_objects.remove(versionKey);
+        emit objectsChanged();
+    }
+}
+
+void DBusObjectRegistry::refreshObjects()
+{
+    ObjectRegistryIPCDBusProxy objectRegistryProxy;
+    objectRegistryProxy.ipc()->setServiceName(m_serviceName);
+    objectRegistryProxy.connectToServer();
+    updateObjects(objectRegistryProxy.getObjects());
+}
+
+const QMap<QString, QString>& DBusObjectRegistry::objects(bool blocking)
 {
     init();
-    if (m_master) {
-        return m_master->objects();
+    if (!isMaster() && blocking && m_registryVersion == invalidRegistryVersion) {
+        refreshObjects();
+    }
+    return m_objects;
+}
+
+void DBusObjectRegistry::onObjectAdded(const QString &objectPath, const QString &serviceName, int registryVersion)
+{
+    if (hasValidObjects() && registryVersion == nextVersion(m_registryVersion)) {
+        m_registryVersion = registryVersion;
+        if (!m_objects.contains(objectPath)) {
+            m_objects.insert(objectPath, serviceName);
+            emit objectsChanged();
+        } else {
+            qCCritical(LogIpc) << "Cannot add object with object path:" << objectPath << ", service name:" << serviceName
+                               << "Object registry already contains this object path with service name:" << m_objects[objectPath];
+        }
+    } else if (registryVersion != version()) {
+        refreshObjects(); // critical situation, intermediate version is missing. Force update
     } else {
-        if (blocking) {
-            if (m_objectRegistryProxy == nullptr) {
-                m_objectRegistryProxy = new facelift::ipc::dbus::ObjectRegistryIPCDBusProxy();
-                m_objectRegistryProxy->ipc()->setServiceName(m_serviceName);
-                m_objectRegistryProxy->connectToServer();
-            }
-            return m_objectRegistryProxy->objects();
+        qCWarning(LogIpc) << "Unexpected object registry with current version. Object path:" << objectPath << ", service name:" << serviceName;
+    }
+}
+
+void DBusObjectRegistry::onObjectRemoved(const QString &objectPath, int registryVersion)
+{
+    if (hasValidObjects() && registryVersion == nextVersion(m_registryVersion)) {
+        m_registryVersion = registryVersion;
+        if (m_objects.remove(objectPath)) {
+            emit objectsChanged();
         } else {
-            return m_objectRegistryAsyncProxy->objects();
+            qCCritical(LogIpc) << "Object does not exist. Object path:" << objectPath;
         }
+    } else if (registryVersion != version()) {
+        refreshObjects(); // critical situation, intermediate version is missing. Force update
```

Comment:
I don't think this is right

Reply:
So what is right then? This should never happen, otherwise we have critical error. This is error handling.